### PR TITLE
Fix issue when creating StarTree segment with large dataset causing JVM crashes

### PR DESCRIPTION
**Pull request by victor2100**
This is an important bug fix since in current version, creating StarTree segment will always cause the JVM crashes when the length * totalSizeInBytes is greater than MAX_INT.

Since in byte representation, MAX_INT can only represent up to 2GB size, which means in order to deal with addresses of even large memory mapped file, we need to use the long type to represent the address space greater than 2GB.

The JVM crash is caused by the overflow of int type, which means when accessing the byte address greater than MAX_INT, current implementation results in a negative number address (integer overflow), causing MAP_ERR.

### CODE REVIEW

**Comment by mayankshriv** [comment]
In general I do not like the approach of declaring a variable as long and then down cast it all over the place to int. Given these values (*sizeInBytes) will fit in int can you still keep them as int, and assign the result of multiplication to long?

File: pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java [hunk 1]
```diff
@@ -15,36 +15,35 @@
...
+ */
 public class StarTreeDataTable {
 
   private File file;
-  private int dimensionSizeInBytes;
-  private int metricSizeInBytes;
-  private int totalSizeInBytes;
+  private long dimensionSizeInBytes;
+  private long metricSizeInBytes;
+  private long totalSizeInBytes;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java]: +1. Make these 3 fields int, but when calculating startOffset, do "long startOffset = startRecordId * (long) totalSizeInBytes"

Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java]: Yep, I updated the code.

**Comment by mayankshriv** [comment]
This is a nit, but can you do something like: (long) (a * b), instead of (a * (long) b)?

**Comment by victor2100** [comment]
(long) (a * b) not working. You can try the following:

int a = Integer.MAX_VALUE;
int b = 1;
System.out.println((long)(a + b)); // -2147483648

File: pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java [hunk 2]
```diff
@@ -211,12 +209,12 @@ public void remove() {
         public Pair<byte[], byte[]> next() {
           byte[] dimBuff = new byte[dimensionSizeInBytes];
           byte[] metBuff = new byte[metricSizeInBytes];
-          mappedByteBuffer.copyTo(pointer * totalSizeInBytes, dimBuff, 0, dimensionSizeInBytes);
-//          mappedByteBuffer.position(pointer * totalSizeInBytes);
-//          mappedByteBuffer.get(dimBuff);
+          copyTo(mappedByteBuffer, pointer * (long) totalSizeInBytes, dimBuff, 0, dimensionSizeInBytes);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java]: The simpler way is "mappedByteBuffer.toDirectByteBuffer(pointer * (long) totalSizeInBytes, dimensionSizeInBytes).get(dimBuff)"

File: pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java [hunk 3]
```diff
@@ -211,12 +209,12 @@ public void remove() {
...
+//        mappedByteBuffer.position(pointer * totalSizeInBytes);
+//        mappedByteBuffer.get(dimBuff);
           if (metricSizeInBytes > 0) {
-            mappedByteBuffer.copyTo(pointer * totalSizeInBytes + dimensionSizeInBytes, metBuff, 0, metricSizeInBytes);
-//            mappedByteBuffer.get(metBuff);
+            copyTo(mappedByteBuffer, pointer * (long) totalSizeInBytes + dimensionSizeInBytes, metBuff, 0, metricSizeInBytes);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java]: Same here.

**Comment by kishoreg** [comment]
High level question. We don't have to change anything to mmap more than 2gb?

File: pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java (hunk 2 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java]: mappedByteBuffer.toDirectByteBuffer(pointer * (long) totalSizeInBytes, dimensionSizeInBytes) results in an array, you still need to iterate the array to get the content. Probably you are checking a newer version of the library?

**Comment by victor2100** [comment]
@kishoreg I think the mmap library should support file greater than 2GB, which is just one of their selling feature. We just need to assure passing the right offset/address (which may be greater than MAX_INT) to it.

**Comment by kishoreg** [comment]
Got it. Looks like feature of larray

File: pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java (hunk 2 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTable.java]: Are you sure "toDirectByteBuffer" returns an array? (not "toDirectByteBuffers", these two are different, and the return type array is used for longer than MAX_INT copy length, which is not necessary for our purpose).

**Pull request closed without merging**